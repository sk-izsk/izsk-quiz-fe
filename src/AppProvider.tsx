import { CssBaseline, ThemeProvider } from '@material-ui/core';
import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { store } from './redux';
import { theme } from './theme/muiTheme';

export interface AppProviderProps {}

const AppProvider: React.FC<AppProviderProps> = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      <Provider store={store}>
        <CssBaseline />
        <BrowserRouter>{children}</BrowserRouter>
      </Provider>
    </ThemeProvider>
  );
};

export { AppProvider };
