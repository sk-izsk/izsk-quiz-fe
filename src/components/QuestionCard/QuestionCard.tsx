import { Box, makeStyles } from '@material-ui/core';
import clsx from 'clsx';
import React, { useState } from 'react';
import { useSelector } from 'react-redux';
import { Body1, Button, CardAction, Divider, H6, Radio, RadioGroup } from 'ui-neumorphism';
import { CardContainer } from '..';
import { axiosAuthorization } from '../../api';
import { QuestionList, Questions } from '../../redux/questionListSlice';
import { RootState } from '../../redux/store';
import { CustomTheme, theme } from '../../theme/muiTheme';
import { getDifficultyOfQuiz, getTypeOfQuiz, header } from '../../utils';
import { ErrorCard } from '../ErrorCard/ErrorCard';
import { LoadingScreen } from '../Loader/Loader';
import { QuizResultModal } from '../QuizResultModal/QuizResultModal';

export interface QuestionCardProps {}

const useStyles = makeStyles((theme: CustomTheme) => ({
  mainContainer: {
    textAlign: 'left',
    maxWidth: 400,
    width: '100%',
  },
  btnContainer: {
    display: 'flex',
    justifyContent: 'flex-end',
  },
  btn: {
    maxWidth: 100,
    width: '100%',
  },
  timerContainer: {
    display: 'flex',
    justifyContent: 'flex-end',
    marginBottom: theme.spacing(1),
  },
  question: {
    marginTop: theme.spacing(1),
    marginBottom: theme.spacing(1),
    display: 'flex',
    flexDirection: 'row',
  },
  typeContainer: {
    display: 'flex',
    justifyContent: 'space-between',
  },
  errorContainer: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
  errorBtn: {
    maxWidth: 200,
    width: '100%',
    marginTop: theme.spacing(1),
  },
  boldText: {
    fontWeight: 'bold',
  },
}));

const QuestionCard: React.FC<QuestionCardProps> = () => {
  const classes = useStyles();
  const [answer, setAnswer] = useState<number | undefined>(undefined);
  const [resultModal, setResultModal] = useState<boolean>(false);
  const [questionNumber, setQuestionNumber] = useState<number>(0);
  const [cardLoading, setCardLoading] = useState<boolean>(false);
  const [disable, setDisable] = useState<boolean>(false);
  const [correct, setCorrect] = useState<number>(0);
  const questions: Questions = useSelector<RootState, Questions>((state: RootState) => state.questions);
  const handleNextQuestion = () => {
    if (questions.questions !== undefined && questions.questions[questionNumber].indexOfCorrectAnswer === answer) {
      setCorrect(correct + 1);
    }
    if (questions.questions?.length === questionNumber + 1) {
      setDisable(true);
      setCardLoading(true);
      setResultModal(true);
      setAnswer(undefined);
    } else {
      setAnswer(undefined);
      setQuestionNumber(questionNumber + 1);
    }
  };

  const handleAnswers = (event: any) => {
    setAnswer(event.value);
  };

  const handleCloseResultModal = async () => {
    await sendQuizHistoryData();
    setResultModal(false);
  };

  const pageReload = async () => {
    await sendQuizHistoryData();
    window.location.reload();
  };

  let quizDetails: any;

  if (questions.questions !== undefined && questions.questions.length > 0) {
    quizDetails = {
      date: new Date(),
      correctAnswer: correct,
      totalQuestion: questions.questions?.length,
      title: getTypeOfQuiz(questions.questions as QuestionList[]) as string,
      type: getDifficultyOfQuiz(questions.questions as QuestionList[]) as string,
    };
  }

  const sendQuizHistoryData = async () => {
    try {
      await axiosAuthorization.put('/quiz-history', quizDetails as any, { headers: header });
    } catch (err) {
      console.warn(err);
    }
  };

  return (
    <>
      <QuizResultModal {...quizDetails} onClose={handleCloseResultModal} retry={pageReload} visible={resultModal} />
      {questions.questions !== undefined && questions.questions.length > 0 ? (
        <Box className={classes.mainContainer}>
          <CardContainer
            cardStyle={classes.mainContainer}
            cardLoading={cardLoading}
            cardAction={
              <CardAction className={classes.btnContainer}>
                <Button
                  disabled={disable || answer === undefined}
                  onClick={handleNextQuestion}
                  className={classes.btn}
                  rounded
                  color={theme.palette.primary.main}
                >
                  Next
                </Button>
              </CardAction>
            }
            inset={true}
          >
            <Box className={classes.typeContainer}>
              <Body1>
                Category: <span className={classes.boldText}>{questions.questions[questionNumber].category}</span>
              </Body1>
              <Body1>
                Type: <span className={classes.boldText}>{questions.questions[questionNumber].type}</span>
              </Body1>
            </Box>
            <Divider style={{ marginTop: theme.spacing(1) }} />
            <H6 className={clsx([classes.question, classes.boldText])}>
              {questionNumber + 1}:
              <div dangerouslySetInnerHTML={{ __html: questions.questions[questionNumber].question }} />
            </H6>
            <Divider elevated />
            <RadioGroup
              disabled={disable}
              vertical
              value={answer}
              color={theme.palette.secondary.main}
              onChange={handleAnswers}
            >
              {questions.questions[questionNumber].answers.map((answer: string, index: number) => {
                return <Radio key={answer} value={index} label={answer} />;
              })}
            </RadioGroup>
          </CardContainer>
        </Box>
      ) : questions.responseCode === 1 ? (
        <Box className={classes.errorContainer}>
          <ErrorCard />
          <Button rounded bordered color={theme.palette.primary.main} onClick={pageReload} className={classes.errorBtn}>
            Create new quiz
          </Button>
        </Box>
      ) : (
        <LoadingScreen />
      )}
    </>
  );
};

export { QuestionCard };
