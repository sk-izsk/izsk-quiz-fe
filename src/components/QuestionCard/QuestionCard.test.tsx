import { render } from '@testing-library/react';
import React from 'react';
import { AppProvider } from '../../AppProvider';
import { QuestionCard } from './QuestionCard';

test('QuestionCard renders correctly', () => {
  const { asFragment } = render(
    <AppProvider>
      <QuestionCard />
    </AppProvider>,
  );
  expect(asFragment()).toMatchSnapshot();
});
