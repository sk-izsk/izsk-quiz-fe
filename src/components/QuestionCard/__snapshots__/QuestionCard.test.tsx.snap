// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionCard renders correctly 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-loaderContainer-19"
  >
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="0"
      class="__234s  _2YVD0 _NijRA "
      id="_00641_"
      role="progressbar"
      style="width: 89px; height: 89px;"
    >
      <div
        class="_1QMv-"
        style="width: 80px; height: 80px; transform: translate3d(-50%, -50%, 0) rotate(0deg);"
      >
        <svg
          class="_2vn3-"
          viewBox="45.714285714285715 45.714285714285715 91.42857142857143 91.42857142857143"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            class="_53Md9"
            cx="91.42857142857143"
            cy="91.42857142857143"
            r="40"
            style="stroke-width: 10; stroke-dasharray: 251.327; stroke-dashoffset: 251.32741228718345;"
          />
        </svg>
      </div>
      <div
        class="_3Jx6J"
        style="width: 85px; height: 85px;"
      />
      <div
        class="_1iTMT"
        style="width: 54px; height: 54px;"
      />
      <label
        class="_2hjq3"
        style="font-size: 24px;"
      />
    </div>
  </div>
</DocumentFragment>
`;
